File: packages/sim-core/src/memory.ts
import type { AgentState, MemoryEvent } from "./types.js";

/**
 * Append a short text event to the agent’s memory, cap length to `maxSlots` (oldest dropped),
 * then randomly remove remaining entries with probability `decayDropProbability` per entry.
 * Not re-exported from the package public entry; used by {@link applyStep}.
 */
export function pushMemory(
  agent: AgentState,
  tick: number,
  summary: string,
  maxSlots: number,
  decayDropProbability: number,
  rnd: () => number,
): void {
  agent.memory.push({ tick, summary });
  while (agent.memory.length > maxSlots) {
    agent.memory.shift();
  }
  if (decayDropProbability > 0) {
    agent.memory = agent.memory.filter(() => rnd() > decayDropProbability);
  }
}

File: packages/ip-sim-engine/src/sim/memory.rs
//! Matches `packages/sim-core/src/memory.ts`.

use crate::sim::types::{Agent, MemoryEvent};
use crate::rng::Rng01;

pub fn push_memory<R: Rng01 + ?Sized>(
    agent: &mut Agent,
    tick: u32,
    summary: &str,
    max_slots: usize,
    decay_drop_probability: f64,
    rnd: &mut R,
) {
    agent.memory.push(MemoryEvent {
        tick,
        summary: summary.to_string(),
    });
    while agent.memory.len() > max_slots {
        agent.memory.remove(0);
    }
    if decay_drop_probability > 0.0 {
        agent.memory.retain(|_| rnd.next_f64() > decay_drop_probability);
    }
}
